Bug 1778958: Remove serviceAccount in favour of serviceAccountName

andrewballantyne opened the pull request:
- Pipeline Operator 0.8.0 deprecates `serviceAccount` in favour of `serviceAccountName`
- Pipeline Operator 0.9.0 will make `serviceAccount` an invalid field

https://jira.coreos.com/browse/ODC-2439

[Tekton Pipeline 0.8.0 Release Notes](https://github.com/tektoncd/pipeline/releases/tag/v0.8.0)

andrewballantyne commented:
/hold

Let's not get this in just yet. More investigation can be done to see the longevity of a Pipeline Operator and see if we can better transition rather than hard cut. 0.7.0 does NOT support the `serviceAccountName` field.

andrewballantyne commented:
Resized this to allow for ports back to 4.3 & 4.2. The YAMLs are not used in todays scripts, so deleting them serves as nothing but potential conflict as I cherry pick to previous releases.

andrewballantyne commented:
Waiting for Canary 0.9.0 before trying to push this through - shouldn't be too much longer. Once it's tested with 0.9.0, we should be good to go with this change & the backports.

andrewballantyne commented on frontend/packages/dev-console/src/utils/pipeline-actions.tsx:
```diff
@@ -57,12 +57,12 @@ export const getPipelineRunData = (
             },
     },
     spec: {
+      ..._.get(latestRun, 'spec', {}),
```
Going with a specific spread to allow wider version support.

andrewballantyne commented on frontend/packages/dev-console/src/utils/pipeline-actions.tsx:
```diff
@@ -44,7 +44,7 @@ export const getPipelineRunData = (
 
   return {
-    apiVersion: `${PipelineRunModel.apiGroup}/${PipelineRunModel.apiVersion}`,
+    apiVersion: pipeline ? pipeline.apiVersion : latestRun.apiVersion,
```
Prep for changing apiVersions (unrelated to this fix though - future assistance)

abhinandan13jan replied:
sorted

abhinandan13jan commented:
/assign

abhinandan13jan commented:
@andrewballantyne This works well. I do see a `hold` label on the PR. Would ltgm it once the hold is removed

abhinandan13jan commented:
/lgtm

abhinandan13jan commented:
/retest

andrewballantyne commented:
The Operator got the canary channel for 0.9.x. Tested it against the migration (from 0.8.x to 0.9.x) and saw there is a chance that the user created their own PipelineRun that has a `serviceAccount` and then the Operator got upgraded; this would cause a UI error preventing re-runs.

Changes since last review:

- Added a migration that changes `serviceAccount` to `serviceAccountName` if it exists
- Added tests for the migration code

andrewballantyne commented:
/hold

andrewballantyne commented:
/king cleanup

andrewballantyne commented:
/kind cleanup

andrewballantyne commented:
/hold cancel
/retest
#3847 has fixed the e2e tests

abhinandan13jan commented on frontend/packages/dev-console/src/utils/pipeline-actions.tsx:
```diff
@@ -18,6 +18,32 @@ export const handlePipelineRunSubmit = (pipelineRun: PipelineRun) => {
+  let newPipelineRun = pipelineRun;
+
+  const serviceAccountPath = 'spec.serviceAccount';
+  if (_.has(newPipelineRun, serviceAccountPath)) {
```
Do we need this `if`, could there be a case, the `if` isn't true but `serviceAccoountName` must be added, in that case the `_.merge` should be put outside the `if`

andrewballantyne replied:
Yes it is important. The goal of this method is to grow with migration paths as tekton evolves. Each if-block will cover a single migration path. 

If you don't need a migration applied the object isn't modified and the weight of this function is next to nothing.

andrewballantyne commented:
/retest

abhinandan13jan commented:
/lgtm

andrewballantyne commented:
/cherry-pick release-4.3

andrewballantyne commented:
/cherry-pick release-4.2
